File: lib/matching/purchase-order.ts
import { supabaseAdmin } from "@/lib/supabase/server";
import type { ControlResult, ExtractedInvoice, PoMatchTier } from "@/lib/types";
import { matchReceipts } from "@/lib/matching/receipt";
import { parseDecimalToCents, centsToDecimalString } from "@/lib/money";
import type { PolicyConfig } from "@/lib/matching/policy";

export type PurchaseOrderMatchResult = {
  purchaseOrderId: string | null;
  tier: PoMatchTier;
  receiptIds: string[];
  controls: ControlResult[];
};

function normalizeDescription(s: string): string {
  return s.toLowerCase().trim();
}

function linesMatch(a: string, b: string): boolean {
  const na = normalizeDescription(a);
  const nb = normalizeDescription(b);
  return na === nb || na.includes(nb) || nb.includes(na);
}

function allowedCentsDiff(baseCents: number, policy: PolicyConfig["unitPriceTolerance"]): number {
  return Math.min(Math.round(baseCents * policy.pct), Math.round(policy.flat * 100));
}

/**
 * PO header + line-item matching, receipt matching, and tolerance
 * evaluation in one pass — CLAUDE.md section 12. No PO number on the
 * invoice is a permitted non-PO category (tier "none", no controls raised);
 * a PO number that doesn't resolve to a real record is a real problem, not
 * a graceful fallback.
 */
export async function matchPurchaseOrder(
  db: ReturnType<typeof supabaseAdmin>,
  extracted: ExtractedInvoice,
  supplierId: string | null,
  policy: PolicyConfig
): Promise<PurchaseOrderMatchResult> {
  const poNumberField = extracted.purchaseOrderNumber;

  if (poNumberField.status !== "verified" || !poNumberField.value) {
    return { purchaseOrderId: null, tier: "none", receiptIds: [], controls: [] };
  }

  const { data: po } = await db.from("purchase_orders").select("*").eq("po_number", poNumberField.value).maybeSingle();

  if (!po) {
    return {
      purchaseOrderId: null,
      tier: "none",
      receiptIds: [],
      controls: [
        {
          controlId: "po_not_found",
          label: "Purchase-order lookup",
          status: "failed",
          severity: "critical",
          reason: `PO number "${poNumberField.value}" was extracted from the invoice, but no matching purchase order exists on file.`,
          evidenceReferences: ["purchaseOrderNumber"],
          blocking: true,
        },
      ],
    };
  }

  const controls: ControlResult[] = [];

  controls.push({
    controlId: "po_status",
    label: "Purchase-order status",
    status: po.status === "open" ? "passed" : "failed",
    severity: po.status === "open" ? "low" : "critical",
    reason:
      po.status === "open"
        ? `${po.po_number} is open.`
        : `${po.po_number} is ${po.status}, not open — closed or cancelled purchase orders block automatic routing.`,
    evidenceReferences: ["purchaseOrderNumber"],
    blocking: true,
  });

  if (supplierId && po.supplier_id !== supplierId) {
    controls.push({
      controlId: "po_supplier_mismatch",
      label: "PO-to-supplier match",
      status: "failed",
      severity: "critical",
      reason: `${po.po_number} is issued to a different supplier than the one matched on this invoice.`,
      evidenceReferences: ["purchaseOrderNumber", "supplierTaxId"],
      blocking: true,
    });
  }

  const { data: poLines } = await db.from("po_lines").select("*").eq("purchase_order_id", po.id).order("line_number");
  const { receiptIds, receivedQuantityByDescription } = await matchReceipts(db, po.id);

  const claimed = new Set<string>();
  const cleanMatches: string[] = [];
  const unmatchedLines: string[] = [];

  for (const line of extracted.lineItems) {
    if (line.description.status !== "verified" || !line.description.value) {
      unmatchedLines.push(`Line ${line.lineNumber}: description not independently verified against the document.`);
      continue;
    }

    const candidate = (poLines ?? []).find((pl) => !claimed.has(pl.id) && linesMatch(line.description.value as string, pl.description));

    if (!candidate) {
      const invoiceTotal = extracted.total.status === "verified" ? parseDecimalToCents(extracted.total.value) : null;
      const notToExceedCents = parseDecimalToCents(String(po.not_to_exceed));
      let overageNote = "";
      if (invoiceTotal !== null && notToExceedCents !== null && invoiceTotal > notToExceedCents) {
        const overageCents = invoiceTotal - notToExceedCents;
        const overagePct = ((overageCents / notToExceedCents) * 100).toFixed(1);
        overageNote = ` This pushes the invoice $${centsToDecimalString(overageCents)} (${overagePct}%) over the PO's not-to-exceed amount of $${po.not_to_exceed}, beyond the lower-of-${(policy.totalInvoiceTolerance.pct * 100).toFixed(0)}%/$${policy.totalInvoiceTolerance.flat} total tolerance.`;
      }
      unmatchedLines.push(
        `"${line.description.value}" ($${line.lineTotal.value ?? "unknown"}) is not on ${po.po_number} and has no receipt or separate authorization.${overageNote}`
      );
      continue;
    }

    claimed.add(candidate.id);

    const lineIssues: string[] = [];

    if (line.unitPrice.status === "verified" && line.unitPrice.value) {
      const invoiceCents = parseDecimalToCents(line.unitPrice.value);
      const poCents = parseDecimalToCents(String(candidate.unit_price));
      if (invoiceCents !== null && poCents !== null) {
        const diff = Math.abs(invoiceCents - poCents);
        const allowed = allowedCentsDiff(poCents, policy.unitPriceTolerance);
        if (diff > allowed) {
          controls.push({
            controlId: "po_unit_price_tolerance",
            label: `Unit-price tolerance — ${candidate.description}`,
            status: "failed",
            severity: "high",
            reason: `Invoiced $${line.unitPrice.value} vs. ${po.po_number} approved $${candidate.unit_price} — $${centsToDecimalString(diff)} over, exceeding the lower-of-${(policy.unitPriceTolerance.pct * 100).toFixed(0)}%/$${policy.unitPriceTolerance.flat} tolerance ($${centsToDecimalString(allowed)} max allowed).`,
            evidenceReferences: ["purchaseOrderNumber"],
            blocking: true,
          });
          lineIssues.push("price");
        }
      }
    }

    if (line.quantity.status === "verified" && line.quantity.value) {
      const invoiceQty = Number(line.quantity.value);
      if (Number.isFinite(invoiceQty) && invoiceQty > candidate.approved_quantity) {
        const receivedQty = receivedQuantityByDescription.get(normalizeDescription(candidate.description)) ?? 0;
        const excessCovered = policy.quantityTolerance.requiresReceiptForExcess ? receivedQty >= invoiceQty : true;
        if (!excessCovered) {
          controls.push({
            controlId: "po_quantity_tolerance",
            label: `Quantity tolerance — ${candidate.description}`,
            status: "failed",
            severity: "high",
            reason: `Invoiced ${invoiceQty} vs. ${po.po_number} approved maximum ${candidate.approved_quantity} — ${invoiceQty - candidate.approved_quantity} over, with no receipt recording the additional quantity. Quantity tolerance is zero without receipt evidence.`,
            evidenceReferences: ["purchaseOrderNumber"],
            blocking: true,
          });
          lineIssues.push("quantity");
        }
      }
    }

    if (lineIssues.length === 0) {
      cleanMatches.push(candidate.description);
    }
  }

  if (cleanMatches.length > 0) {
    controls.push({
      controlId: "po_line_match",
      label: "Matched PO line(s)",
      status: "passed",
      severity: "low",
      reason: `${cleanMatches.length} line(s) match ${po.po_number} within tolerance: ${cleanMatches.join(", ")}.`,
      evidenceReferences: ["purchaseOrderNumber"],
      blocking: false,
    });
  }

  for (const note of unmatchedLines) {
    controls.push({
      controlId: "po_unmatched_line",
      label: "Unmatched invoice line",
      status: "failed",
      severity: "high",
      reason: note,
      evidenceReferences: ["purchaseOrderNumber"],
      blocking: true,
    });
  }

  const hasProblems = controls.some((c) => c.status === "failed" && c.controlId !== "po_status") || unmatchedLines.length > 0;
  let tier: PoMatchTier;
  if (unmatchedLines.length === extracted.lineItems.length) {
    tier = "ambiguous";
  } else if (!hasProblems) {
    tier = "exact";
  } else {
    tier = "partial";
  }

  return { purchaseOrderId: po.id, tier, receiptIds, controls };
}

File: ledgerguard/matching/purchase_order.py
"""PO header + line-item matching, receipt matching, and tolerance
evaluation in one pass."""

from sqlalchemy import select
from sqlalchemy.orm import Session

from ..models import PoLine, PurchaseOrder
from ..money import cents_to_decimal_string, parse_decimal_to_cents
from ..policy import PolicyConfig
from .receipt import match_receipts


def _normalize_description(s: str) -> str:
    return s.lower().strip()


def _lines_match(a: str, b: str) -> bool:
    na, nb = _normalize_description(a), _normalize_description(b)
    return na == nb or nb in na or na in nb


def _allowed_cents_diff(base_cents: int, tolerance) -> int:
    return min(round(base_cents * tolerance.pct), round(tolerance.flat * 100))


def match_purchase_order(db: Session, extracted: dict, supplier_id: str | None, policy: PolicyConfig) -> dict:
    """No PO number on the invoice is a permitted non-PO category (tier
    "none", no controls raised); a PO number that doesn't resolve to a real
    record is a real problem, not a graceful fallback.

    Returns {"purchase_order_id", "tier", "receipt_ids", "controls"}.
    """
    po_number_field = extracted["purchaseOrderNumber"]

    if po_number_field.status != "verified" or not po_number_field.value:
        return {"purchase_order_id": None, "tier": "none", "receipt_ids": [], "controls": []}

    po = db.scalar(select(PurchaseOrder).where(PurchaseOrder.po_number == po_number_field.value))

    if not po:
        return {
            "purchase_order_id": None,
            "tier": "none",
            "receipt_ids": [],
            "controls": [
                {
                    "controlId": "po_not_found",
                    "label": "Purchase-order lookup",
                    "status": "failed",
                    "severity": "critical",
                    "reason": f'PO number "{po_number_field.value}" was extracted from the invoice, but no matching purchase order exists on file.',
                    "evidenceReferences": ["purchaseOrderNumber"],
                    "blocking": True,
                }
            ],
        }

    controls: list[dict] = []

    controls.append(
        {
            "controlId": "po_status",
            "label": "Purchase-order status",
            "status": "passed" if po.status == "open" else "failed",
            "severity": "low" if po.status == "open" else "critical",
            "reason": (
                f"{po.po_number} is open."
                if po.status == "open"
                else f"{po.po_number} is {po.status}, not open — closed or cancelled purchase orders block automatic routing."
            ),
            "evidenceReferences": ["purchaseOrderNumber"],
            "blocking": True,
        }
    )

    if supplier_id and po.supplier_id != supplier_id:
        controls.append(
            {
                "controlId": "po_supplier_mismatch",
                "label": "PO-to-supplier match",
                "status": "failed",
                "severity": "critical",
                "reason": f"{po.po_number} is issued to a different supplier than the one matched on this invoice.",
                "evidenceReferences": ["purchaseOrderNumber", "supplierTaxId"],
                "blocking": True,
            }
        )

    po_lines = db.scalars(select(PoLine).where(PoLine.purchase_order_id == po.id).order_by(PoLine.line_number)).all()
    receipt_match = match_receipts(db, po.id)
    received_by_desc = receipt_match["received_quantity_by_description"]

    claimed: set = set()
    clean_matches: list[str] = []
    unmatched_lines: list[str] = []

    for line in extracted["lineItems"]:
        if line["description"].status != "verified" or not line["description"].value:
            unmatched_lines.append(f"Line {line['lineNumber']}: description not independently verified against the document.")
            continue

        candidate = next(
            (pl for pl in po_lines if pl.id not in claimed and _lines_match(line["description"].value, pl.description)),
            None,
        )

        if not candidate:
            invoice_total = parse_decimal_to_cents(extracted["total"].value) if extracted["total"].status == "verified" else None
            not_to_exceed_cents = parse_decimal_to_cents(str(po.not_to_exceed))
            overage_note = ""
            if invoice_total is not None and not_to_exceed_cents is not None and invoice_total > not_to_exceed_cents:
                overage_cents = invoice_total - not_to_exceed_cents
                overage_pct = round((overage_cents / not_to_exceed_cents) * 100, 1)
                overage_note = (
                    f" This pushes the invoice ${cents_to_decimal_string(overage_cents)} ({overage_pct}%) over "
                    f"the PO's not-to-exceed amount of ${po.not_to_exceed}, beyond the "
                    f"lower-of-{round(policy.total_invoice_tolerance.pct * 100)}%/${policy.total_invoice_tolerance.flat:g} total tolerance."
                )
            line_total_value = line["lineTotal"].value or "unknown"
            unmatched_lines.append(
                f"\"{line['description'].value}\" (${line_total_value}) is not on {po.po_number} and has no "
                f"receipt or separate authorization.{overage_note}"
            )
            continue

        claimed.add(candidate.id)
        line_issues: list[str] = []

        if line["unitPrice"].status == "verified" and line["unitPrice"].value:
            invoice_cents = parse_decimal_to_cents(line["unitPrice"].value)
            po_cents = parse_decimal_to_cents(str(candidate.unit_price))
            if invoice_cents is not None and po_cents is not None:
                diff = abs(invoice_cents - po_cents)
                allowed = _allowed_cents_diff(po_cents, policy.unit_price_tolerance)
                if diff > allowed:
                    controls.append(
                        {
                            "controlId": "po_unit_price_tolerance",
                            "label": f"Unit-price tolerance — {candidate.description}",
                            "status": "failed",
                            "severity": "high",
                            "reason": (
                                f"Invoiced ${line['unitPrice'].value} vs. {po.po_number} approved "
                                f"${candidate.unit_price} — ${cents_to_decimal_string(diff)} over, exceeding "
                                f"the lower-of-{round(policy.unit_price_tolerance.pct * 100)}%/"
                                f"${policy.unit_price_tolerance.flat:g} tolerance "
                                f"(${cents_to_decimal_string(allowed)} max allowed)."
                            ),
                            "evidenceReferences": ["purchaseOrderNumber"],
                            "blocking": True,
                        }
                    )
                    line_issues.append("price")

        if line["quantity"].status == "verified" and line["quantity"].value:
            try:
                invoice_qty = float(line["quantity"].value)
            except ValueError:
                invoice_qty = None
            if invoice_qty is not None and invoice_qty > float(candidate.approved_quantity):
                received_qty = received_by_desc.get(_normalize_description(candidate.description), 0)
                excess_covered = received_qty >= invoice_qty if policy.quantity_tolerance.requires_receipt_for_excess else True
                if not excess_covered:
                    controls.append(
                        {
                            "controlId": "po_quantity_tolerance",
                            "label": f"Quantity tolerance — {candidate.description}",
                            "status": "failed",
                            "severity": "high",
                            "reason": (
                                f"Invoiced {invoice_qty:g} vs. {po.po_number} approved maximum "
                                f"{float(candidate.approved_quantity):g} — {invoice_qty - float(candidate.approved_quantity):g} "
                                "over, with no receipt recording the additional quantity. Quantity tolerance "
                                "is zero without receipt evidence."
                            ),
                            "evidenceReferences": ["purchaseOrderNumber"],
                            "blocking": True,
                        }
                    )
                    line_issues.append("quantity")

        if not line_issues:
            clean_matches.append(candidate.description)

    if clean_matches:
        controls.append(
            {
                "controlId": "po_line_match",
                "label": "Matched PO line(s)",
                "status": "passed",
                "severity": "low",
                "reason": f"{len(clean_matches)} line(s) match {po.po_number} within tolerance: {', '.join(clean_matches)}.",
                "evidenceReferences": ["purchaseOrderNumber"],
                "blocking": False,
            }
        )

    for note in unmatched_lines:
        controls.append(
            {
                "controlId": "po_unmatched_line",
                "label": "Unmatched invoice line",
                "status": "failed",
                "severity": "high",
                "reason": note,
                "evidenceReferences": ["purchaseOrderNumber"],
                "blocking": True,
            }
        )

    has_problems = any(c["status"] == "failed" and c["controlId"] != "po_status" for c in controls) or bool(unmatched_lines)
    if len(unmatched_lines) == len(extracted["lineItems"]):
        tier = "ambiguous"
    elif not has_problems:
        tier = "exact"
    else:
        tier = "partial"

    return {"purchase_order_id": po.id, "tier": tier, "receipt_ids": receipt_match["receipt_ids"], "controls": controls}
